h2o-k8s/README.md

```diff
@@ -0,0 +1,103 @@
+on a Kubernetes cluster. Kubernetes assumes all the pods inside the cluster are stateful and does not attempt to restart
+the individual pods on failure. Once a job is triggered on an H2O cluster, the cluster is locked and no additional nodes
+can be added. Therefore, the cluster has to be restarted as a whole if required - which is a perfect fit for a StatefulSet.
+ 
+
+```yaml
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: h2o-stateful-set
+  namespace: h2o-statefulset
+spec:
+  serviceName: h2o-service
+  replicas: 3
+  selector:
+    matchLabels:
+      app: h2o-k8s
+  template:
+    metadata:
+      labels:
+        app: h2o-k8s
+    spec:
+      terminationGracePeriodSeconds: 10
+      containers:
+        - name: h2o-k8s
+          image: 'pscheidl/h2o-k8s'
+          ports:
+            - containerPort: 54321
+              protocol: TCP
+          env:
+          - name: H2O_KUBERNETES_SERVICE_DNS
+            value: h2o-service.h2o-statefulset.svc.cluster.local
+          - name: H2O_NODE_LOOKUP_TIMEOUT
+            value: '180'
+          - name: H2O_NODE_EXPECTED_COUNT
+            value: '3'
+```
+Besides standardized Kubernetes settings, like `replicas: 3` defining the number of pods with H2O instantiated, there are
+several settings to pay attention to.
+
+The name of the application `app: h2o-k8s` must correspond to the name expected by the above-defined headless service in order
+for the H2O node discovery to work. H2O communicates on port 54321, therefore `containerPort: 54321`must be exposed to
+make it possible for the clients to connect.
+
+Environment variables:
+
+1. `H2O_KUBERNETES_SERVICE_DNS` - **[MANDATORY]** Crucial for the clustering to work. The format usually follows the
+ `<service-name>.<project-name>.svc.cluster.local` pattern. This setting enables H2O node discovery via DNS.
+  It must be modified to match the name of the headless service created. Also, pay attention to the rest of the address
+  to match the specifics of your Kubernetes implementation.
+1. `H2O_NODE_LOOKUP_TIMEOUT` - **[OPTIONAL]** Node lookup constraint. Time before the node lookup is ended. 
+1. `H2O_NODE_EXPECTED_COUNT` - **[OPTIONAL]** Node lookup constraint. Expected number of H2O pods to be discovered.
+
+If none of the optional lookup constraints is specified, a sensible default node lookup timeout will be set - currently
+defaults to 3 minutes. If any of the lookup constraints is defined, the H2O node lookup is terminated on whichever 
+condition is met first.
+
+### Exposing H2O cluster
+
+Exposing the H2O cluster is a responsibility of the Kubernetes administrator. By default, an
+ [Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/) can be created. Different platforms offer
+ different capabilities, e.g. OpenShift offers [Routess](https://docs.openshift.com/container-platform/4.3/networking/routes/route-configuration.html).
```

Comment:
I couldn't find in the documentation a bit where actually h2o is started, am I missing something?

Reply:
Thanks, I'll add more explanation. H2O is started as soon as all pods cluster. Which means the headless service is created and the H2O pods are started. The ingress only makes H2O accessible from the outside. It may even change without affecting the underlying layers.

You may spawn H2O in one go by doing `kubectl apply -f h2o.yaml`. The YAML looks like this and it will work on your local minikube as well: 

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: h2o-k8s
  labels:
    app: h2o-k8s
  namespace: default
spec:
  containers:
    - name: h2o-k8s
      image: pscheidl/h2o-k8s
      ports:
        - containerPort: 54321
      env:
      - name: H2O_KUBERNETES_SERVICE_DNS
        value: h2o-service.default.svc.cluster.local
---
apiVersion: v1
kind: Service
metadata:
  name: h2o-service
spec:
  type: ClusterIP
  clusterIP: None
  selector:
    app: h2o-k8s
  ports:
  - protocol: TCP
    port: 54321
```

Reply:
Nice!

Reply:
@Pscheidl `Routess` looks like a typo

Reply:
Fixed, thanks.